### PR TITLE
Miscellaneous speedups for IRC/LS (CFG)

### PR DESCRIPTION
**Pull request by xclerc**
(On top of https://github.com/ocaml-flambda/flambda-backend/pull/1007.)

The commits (starting from https://github.com/ocaml-flambda/flambda-backend/commit/55bb87b80c4bc9860b6e1ecc94c7aac58ed76fe3) are, to the best
of my knowledge, independent so that we can pick
and choose.

I have benchmarked at various steps on the build of
the compiler. This is of course not the greatest of
benchmarks but gives an idea of the speedups we
can expect. Each configuration below was run at least
4 times, removing obvious outliers (about 1 in 30 runs
was significantly slower). Here are the results:

|               | closure | flambda1 | flambda2 |
|:------------- | -------:| --------:| --------:|
| base line [1] |     57s |      80s |     100s |
| [1] with IRC  |     63s |      90s |     111s |
| [1] with LS   |     60s |      85s |     105s |
| [2] with IRC  |     61s |      87s |     106s |
| [2] with LS   |     61s |      87s |     106s |
| [3] with IRC  |     59s |      83s |     103s |
| [3] with LS   |     59s |      82s |     103s |

[1] 39a1c941fb8322b890e67a6c3ca8b534ef5b11bc
[2] d36fb270d6d01de7947c3e128c29c6aab5ece1e9
[3] 44c276d2852e3173d522a51f3453b0d2d8b1e4e0

"with IRC" is
```
$ export BUILD_OCAMLPARAM='_,w=-46,regalloc=irc,regalloc-param=IRC_SPLIT:off,regalloc-param=IRC_SPILLING_HEURISTICS:flat-uses'
$ export OCAMLPARAM='_,w=-46,regalloc=irc,regalloc-param=IRC_SPLIT:off,regalloc-param=IRC_SPILLING_HEURISTICS:flat-uses'
```

"with LS" is
```
$ export BUILD_OCAMLPARAM='_,w=-46,regalloc=ls,regalloc-param=LS_ORDER:layout'
$ export OCAMLPARAM='_,w=-46,regalloc=ls,regalloc-param=LS_ORDER:layout'
```

### CODE REVIEW

**Comment by xclerc** [comment]
(https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/456129c6a277ba43c7386bdbfb33da7f7dfeb72b is significantly easier to review [with whitespace hidden](https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/456129c6a277ba43c7386bdbfb33da7f7dfeb72b?diff=split&w=1).)

**Comment by Gbury** [comment]
I'm a bit confused by the numbers: what's the baseline (i.e. does it corresponds to `main`) ? and why do all of the changes seem to increase the time compared to the baseline ? Basically, I don't understand the numbers as presented because they make it seem (at least to me) as if the various changes actually increase compilation time compared to the baseline (which is the inverse of what i'd expect from speedups).

**Comment by xclerc** [comment]
The baseline is with upstream's pipeline while
the other lines are with the CFG pipeline. An
example of speedup would be to move from
"[1] with IRC" to "[3] with IRC".

**Review by gretay-js — COMMENTED** [review]
The following commits look fine to me.
[d36fb27](https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/d36fb270d6d01de7947c3e128c29c6aab5ece1e9)
[55bb87b](https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/55bb87b80c4bc9860b6e1ecc94c7aac58ed76fe3) 
[0567fbe](https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/0567fbee08fc1e597673da52e655e782f5b3d4cf)
[44c276d](https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/44c276d2852e3173d522a51f3453b0d2d8b1e4e0)
[456129c](https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/456129c6a277ba43c7386bdbfb33da7f7dfeb72b)

**Comment by xclerc** [comment]
Does it mean you dislike https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/456129c6a277ba43c7386bdbfb33da7f7dfeb72b?

**Comment by xclerc** [comment]
At the same time you have [d36fb27](https://github.com/ocaml-flambda/flambda-backend/pull/1290/commits/d36fb270d6d01de7947c3e128c29c6aab5ece1e9)
twice - copy/paste error?

**Review by mshinwell — APPROVED** [review]
Only files owned by me

File: backend/cfg/cfg_ls_utils.mli [hunk 1]
```diff
@@ -34,7 +34,7 @@ val iter_cfg_order :
   Cfg_with_layout.t -> Order.t -> f:(Cfg.basic_block -> unit) -> unit
 
 (* The [trap_handler] parameter to the [instruction] and [terminator] functions
-   is set to [true] iff the instruction is the first one of a block which is a
+   is set to [true] iff the instruction if the first one of a block which is a
```
Comment (gretay-js) [inline backend/cfg/cfg_ls_utils.mli:37]:
```suggestion
   is set to [true] iff the instruction is the first one of a block which is a
```

File: backend/cfg/tests/check_regalloc_validation.ml [hunk 2]
```diff
@@ -367,6 +367,10 @@ let check name f ~exp_std ~exp_err =
     with_wrap_ppf Format.std_formatter (fun () ->
         with_wrap_ppf Format.err_formatter (fun () ->
             try
+              let old_regalloc_validate =
+                !Flambda_backend_flags.regalloc_validate
+              in
+              Flambda_backend_flags.regalloc_validate := true;
```
Comment (gretay-js) [inline backend/cfg/tests/check_regalloc_validation.ml:373]:
```suggestion
```

**Review by gretay-js — APPROVED** [review]

**Pull request merged**